### PR TITLE
Honzon Predeploy Contract

### PR DESCRIPTION
**Pull request by ferrell-code**
companion to: https://github.com/AcalaNetwork/Acala/pull/2143

### CODE REVIEW

File: contracts/honzon/Honzon.sol [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+     * @dev Adjust CDP position
+     * Returns a boolean value indicating whether the operation succeeded.
+     */
+    function adjustLoan(address currencyId, int256 collateralAdjustment, int256 debitAdjustment)
+    public
+    override
+    returns (bool) {
+        require(collateralAdjustment != 0 && debitAdjustment != 0, "Honzon: adjustment amounts are zero");
+
+        (bool success, bytes memory returnData) = precompile.staticcall(abi.encodeWithSignature("adjustLoan(address,address,int256,int256)", msg.sender, currencyId, collateralAdjustment, debitAdjustment));
```
Comment (ermalkaleci) [inline contracts/honzon/Honzon.sol:20]: @ferrell-code staticcall is for executing something without applying changes, like read or view. In this case it needs to be .call(

File: contracts/honzon/Honzon.sol [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+    /**
+     * @dev Close CDP position using DEX
+     * Returns a boolean value indicating whether the operation succeeded.
+     */
+    function closeLoanByDex(address currencyId, uint256 maxCollateralAmount)
+    public
+    override
+    returns (bool) {
+        (bool success, bytes memory returnData) = precompile.staticcall(abi.encodeWithSignature("closeLoanByDex(address,address,uint256)", msg.sender, currencyId, maxCollateralAmount));
```
Comment (ermalkaleci) [inline contracts/honzon/Honzon.sol:39]: same here

File: contracts/honzon/Honzon.sol [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+import "./IHonzon.sol";
+
+contract Honzon is IHonzon {
+    address constant private precompile = address(0x0000000000000000000000000000000000000409);
+
+    /**
+     * @dev Adjust CDP position
+     * Returns a boolean value indicating whether the operation succeeded.
+     */
+    function adjustLoan(address currencyId, int256 collateralAdjustment, int256 debitAdjustment)
```
Comment (ermalkaleci) [inline contracts/honzon/Honzon.sol:14]:
```suggestion
    function adjustLoan(address currencyId, int128 collateralAdjustment, int128 debitAdjustment)
```

Comment (ermalkaleci) [reply contracts/honzon/Honzon.sol:14]: @zjb0807 ^

**Review by zjb0807 — APPROVED** [review]

**Pull request merged**